core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -3034,6 +3038,71 @@ private static class VariableCollector extends RexVisitorImpl<Void> {
     }
   }
 
+  private static final Set<SqlKind> CONSTANT_VALUE_CONSTRUCTOR_KINDS =
+      EnumSet.of(
+          SqlKind.ARRAY_VALUE_CONSTRUCTOR,
+          SqlKind.MAP_VALUE_CONSTRUCTOR,
+          SqlKind.MULTISET_VALUE_CONSTRUCTOR);
+
+  private static @Nullable Pair<RexNode, RexNode> constantEquality(RexCall call) {
+    final RexNode o0 = call.getOperands().get(0);
+    final RexNode o1 = call.getOperands().get(1);
+    if (RexUtil.isReferenceOrAccess(o0, true) && isConstant(o1)) {
+      return Pair.of(o0, o1);
+    }
+    if (RexUtil.isReferenceOrAccess(o1, true) && isConstant(o0)) {
+      return Pair.of(o1, o0);
+    }
+    return null;
+  }
+
+  private static boolean constantsEquivalent(RexNode node1, RexNode node2) {
+    if (Objects.equals(node1, node2)) {
+      return true;
+    }
+    if (!(node1 instanceof RexCall) || !(node2 instanceof RexCall)) {
+      return false;
+    }
+    final RexCall call1 = (RexCall) node1;
+    final RexCall call2 = (RexCall) node2;
+    if (call1.getKind() != call2.getKind()) {
+      return false;
+    }
+    switch (call1.getKind()) {
+    case MULTISET_VALUE_CONSTRUCTOR:
+      return multisetLiteralEquals(call1, call2);
+    case MAP_VALUE_CONSTRUCTOR:
+      return mapLiteralEquals(call1, call2);
+    default:
+      return false;
+    }
+  }
+
+  private static boolean multisetLiteralEquals(RexCall left, RexCall right) {
+    return HashMultiset.create(left.getOperands())
+        .equals(HashMultiset.create(right.getOperands()));
```

Comment:
we just had a conversation about comparing maps for ordering, the rule that you implemented there is not compatible with this one: this can claim equality while ordering can claim inequality.

The question is: are maps ordered or not? I think the answer depends on the dialect.

Reply:
Yes, we just discussed. The current goal is to implement a simplification, but it doesn’t seem very logical to skip sorting. Logically speaking, should we also add sorting in the previous PR? Or should we also avoid sorting in the current PR? If it depends on the dialect, that kind of calculation logic doesn’t seem very straightforward to implement. Would it be better to add sorting to the map in the previous PR?